### PR TITLE
[RyuJit/ARM32] Forbid integer MOD/UMOD/DIV/UDIV after morph.

### PR DESCRIPTION
**Pull request by sandreenko**
Legacy backend has `fgIsSignedDivOptimizable` that is used in codegenlegacy to optimize `DIV % (2^n)`, RyuJit tries to do it as well (in lowering?! `fgIsSignedDivOptimizable`), but it is not that clever, so it can't optimize for example `0 % 1` and that was a problem.

The fix forbids DIV/UDIV/MOD/UMOD integer nodes after morph.

Fix DevDiv_591260, 591210, 591207.

As usual PR is separated on commits. Please review them.

### CODE REVIEW

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT arm Cross Checked r2r Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r_jitminopts Build and Test

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r

**Comment by sandreenko** [comment]
Test case:
after morph we have:
```
[000058] -----+-------             |  |     |  /--*  CNS_INT   int    1
[000059] --CX-+-------             |  |     \--*  MOD       int
[000054] --CX-+-------             |  |        |  /--*  CAST_ovfl int <- ubyte <- int
[000151] --CX-+-------             |  |        |  |  \--*  CALL help int    HELPER.CORINFO_HELP_DBL2INT_OVF
[000053] -----+------- arg0 f0,f1  |  |        |  |     \--*  LCL_VAR   double V13 loc9
[000055] --CX-+-------             |  |        \--*  MUL       int
[000050] -----+-------             |  |           |     /--*  LCL_VAR   double V13 loc9
[000051] N----+---U---             |  |           |  /--*  GT        int
[000049] -----+-------             |  |           |  |  \--*  CAST      double <- int
[000048] -----+-------             |  |           |  |     \--*  NEG       int
[000047] -----+-------             |  |           |  |        \--*  LCL_VAR   int    V17 loc13
[000052] -----+-------             |  |           \--*  XOR       int
[000046] -----+-------             |  |              \--*  CAST      int <- short <- int
[000150] -----+-------             |  |                 \--*  CAST      int <- float
[000045] -----+-------             |  |                    \--*  LCL_VAR   float  V18 loc14
```
`000058` is 2^0, so `fgIsSignedDivOptimizable` returns `true`.

Then IR Rationalize finds that `[000052] == 0`:
```
N013 (  1,  1) [000052] -------------       t52 =    CNS_INT   int    0 $40
N015 (  1,  1) [000058] -------------       t58 =    CNS_INT   int    1 $41
                                                  /--*  t52    int
                                                  +--*  t58    int
N016 ( 28, 23) [000059] ---X---------       t59 = *  MOD       int    $40c
```
and `LowerConstIntDivOrMod` gives up, because it doesn't optimize cases that throws.

**Comment by mikedn** [comment]
Eh, too bad that I haven't finished #15690 yet, this "forbids DIV/UDIV/MOD/UMOD integer nodes after morph" is really the wrong way to do this.

**Comment by mikedn** [comment]
> and LowerConstIntDivOrMod gives up, because it doesn't optimize cases that throws.

To be clear, `0 % 1` does not throw, it produces 0. `LowerConstIntDivOrMod` doesn't take care of this case because it's extremely rare to end up with a node with constant operands here, constant folding should normally take care of this and on x86 it doesn't matter anyway since it supports DIV/MOD. 

It would pretty simple to extend `LowerConstIntDivOrMod` to handle this case but then I'm not sure what other weird cases there are. This kind of decision/transform split between phases is IMO inherently unsafe and getting rid of `fgIsSignedDivOptimizable` is good. 

Unfortunately this means that we won't optimize division by constant on ARM32 and instead we'll always generate helper calls, even for power by 2 division. It looks to me that the only way out of this is to move helper call generation to lowering but it's too late to do that in 2.1.

**Comment by sandreenko** [comment]
> It would pretty simple to extend LowerConstIntDivOrMod to handle this case but then I'm not sure what other weird cases there are. This kind of decision/transform split between phases is IMO inherently unsafe and getting rid of fgIsSignedDivOptimizable is good.

There are other strange cases, so you are right: we do not want to go this way right now.

> Unfortunately this means that we won't optimize division by constant on ARM32 and instead we'll always generate helper calls, even for power by 2 division. It looks to me that the only way out of this is to move helper call generation to lowering but it's too late to do that in 2.1.

I think arm performance won't notice it ☺️  There are too many problems already.

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT arm Cross Checked r2r Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r_jitminopts Build and Test
@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r

**Comment by sandreenko** [comment]
@dotnet-bot test Windows_NT arm Cross Checked r2r Build and Test
@dotnet-bot test Windows_NT arm Cross Checked r2r_jitminopts Build and Test
@dotnet-bot test Windows_NT x86_arm_altjit Checked r2r

**Comment by sandreenko** [comment]
PTAL @dotnet/jit-contrib

**Comment by mikedn** [comment]
It's interesting to note that on x86 the repro does not produce a MOD with constant operands:
```
N018 ( 47, 34) [000055] ---X--------        t55 = *  MUL       int    $38a
N019 (  1,  1) [000058] ------------        t58 =    CNS_INT   int    1 $41
                                                 /--*  t55    int    
                                                 +--*  t58    int    
N020 ( 49, 36) [000059] ---X--------        t59 = *  MOD       int    $38c
```
Somewhere along the way the ARM JIT managed to turn that MUL into a 0.

**Comment by sandreenko** [comment]
> It's interesting to note that on x86 the repro does not produce a MOD with constant operands:
> N018 ( 47, 34) [000055] ---X--------        t55 = *  MUL       int    $38a
> N019 (  1,  1) [000058] ------------        t58 =    CNS_INT   int    1 $41
>                                                  /--*  t55    int    
>                                                  +--*  t58    int    
> N020 ( 49, 36) [000059] ---X--------        t59 = *  MOD       int    $38c
> Somewhere along the way the ARM JIT managed to turn that MUL into a 0.

I think It is something along x86 Jit managed to prevent this optimization:

x86 Trees at start of optValnumCSE_Heuristic():
```
N019 ( 16, 16)              [000046] ------------              |  |        |     /--*  CAST      int <- short <- int $387
N018 ( 15, 14)              [000145] ------------              |  |        |     |  \--*  CAST      int <- double $386
N017 (  9,  8) CSE #02 (def)[000144] ------------              |  |        |     |     \--*  CAST      double <- float $480
N016 (  3,  2)              [000045] ------------              |  |        |     |        \--*  LCL_VAR   float  V18 loc14        u:2 $1c0
```

Promoting CSE:
optValnumCSE morphed tree:
```
N023 ( 23, 21)              [000046] -A----------              |  |        |     /--*  CAST      int <- short <- int $387
N022 ( 22, 19)              [000145] -A----------              |  |        |     |  \--*  CAST      int <- double $386
N020 (  3,  2)              [000180] ------------              |  |        |     |     |  /--*  LCL_VAR   double V25 cse3          $480
N021 ( 16, 13)              [000181] -A----------              |  |        |     |     \--*  COMMA     double $480
N017 (  9,  8)              [000144] ------------              |  |        |     |        |  /--*  CAST      double <- float $480
N016 (  3,  2)              [000045] ------------              |  |        |     |        |  |  \--*  LCL_VAR   float  V18 loc14        u:2 $1c0
N019 ( 13, 11)              [000179] -A------R---              |  |        |     |        \--*  ASG       double $VN.Void
N018 (  3,  2)              [000178] D------N----              |  |        |     |           \--*  LCL_VAR   double V25 cse3          $480
```


so `Promoting CSE:` creates a comma and makes it is harder to optimize.

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -11834,7 +11834,11 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             }
 
 #if USE_HELPERS_FOR_INT_DIV
-            if (typ == TYP_INT && !fgIsSignedDivOptimizable(op2))
+            if (typ == TYP_INT
+#if defined(LEGACY_BACKEND)
+                && !fgIsSignedDivOptimizable(op2)
```
Comment (briansull) [inline src/jit/morph.cpp]: What system do we have that uses HELPERS for INT_DIV but is not a LEGACY_BACKEND?
RyuJIT Arm32 I would guess.

Comment (sandreenko) [reply src/jit/morph.cpp]: Yes, `USE_HELPERS_FOR_INT_DIV` is defined only for arm.

**Comment by sandreenko** [comment]
Thanks, merged.

**Pull request merged**